File: tossquant_cli/hud.py
from __future__ import annotations

import os
import shlex
import shutil
import subprocess
import sys
import time
from pathlib import Path

from .runtime import build_runtime_snapshot, read_runtime_snapshot, record_runtime, render_runtime_line, write_runtime_snapshot

RESET = "\033[0m"
HUD_COLOR = "\033[2m\033[94m"
DEFAULT_TMUX_SESSION = "tossquant"


def color(text: str, ansi: str) -> str:
    return f"{ansi}{text}{RESET}"


def print_hud_once(*, base: str | Path | None = "data", mode: str = "quant", last_action: str = "ready") -> str:
    snapshot = record_runtime(mode=mode, last_action=last_action, base=base)
    line = render_runtime_line(snapshot)
    print(color(line, HUD_COLOR))
    return line


def watch_hud(*, base: str | Path | None = "data", interval: float = 1.0) -> int:
    while True:
        snapshot = read_runtime_snapshot(base)
        if snapshot is None:
            snapshot = build_runtime_snapshot(base=base)
            write_runtime_snapshot(snapshot, base)
        line = render_runtime_line(snapshot)
        print("\033[2J\033[H" + color(line, HUD_COLOR), flush=True)
        time.sleep(max(interval, 0.2))


def tmux_path() -> str | None:
    return shutil.which("tmux")


def tmux_install_hint() -> str:
    return "install tmux with your OS package manager, e.g. apt install tmux, brew install tmux, or pacman -S tmux"


def in_tmux() -> bool:
    return bool(os.environ.get("TMUX"))


def shell_command(parts: list[str]) -> str:
    return " ".join(shlex.quote(str(part)) for part in parts)


def hud_watch_command(*, base: str | Path | None = "data", interval: float = 1.0) -> str:
    return shell_command(
        [
            sys.executable,
            "-m",
            "tossquant_cli.cli",
            "--data-dir",
            str(base or "data"),
            "hud",
            "--watch",
            "--interval",
            str(interval),
        ]
    )


def interactive_command() -> str:
    return shell_command([sys.executable, "-m", "tossquant_cli.cli", "--no-tmux"])


def launch_tmux_hud(*, base: str | Path | None = "data", height: int = 3, interval: float = 1.0, cwd: str | Path | None = None) -> tuple[int, str]:
    tmux = tmux_path()
    if not tmux:
        return 127, f"tmux not found in PATH; {tmux_install_hint()}"
    if not in_tmux():
        return 2, "not inside a tmux session; run tossquant with no arguments or start tmux first"

    height = max(1, int(height))
    command = hud_watch_command(base=base, interval=interval)
    args = [tmux, "split-window", "-v", "-l", str(height)]
    if cwd is not None:
        args.extend(["-c", str(cwd)])
    args.append(command)
    completed = subprocess.run(args, text=True, capture_output=True, check=False)
    message = (completed.stderr or completed.stdout or "tmux HUD launched").strip()
    return int(completed.returncode), message


def launch_tmux_runtime(
    *,
    base: str | Path | None = "data",
    session: str = DEFAULT_TMUX_SESSION,
    height: int = 3,
    interval: float = 1.0,
    cwd: str | Path | None = None,
) -> tuple[int, str]:
    tmux = tmux_path()
    if not tmux:
        return 127, f"tmux not found in PATH; {tmux_install_hint()}"
    if in_tmux():
        return 2, "already inside tmux; use /hud tmux to add the TossQuant HUD pane"

    cwd_path = Path(cwd or Path.cwd())
    height = max(1, int(height))
    target = f"{session}:main"
    main_command = interactive_command()
    hud_command = hud_watch_command(base=base, interval=interval)

    create = subprocess.run(
        [tmux, "new-session", "-d", "-s", session, "-n", "main", "-c", str(cwd_path), main_command],
        text=True,
        capture_output=True,
        check=False,
    )
    if create.returncode != 0:
        existing = subprocess.run([tmux, "has-session", "-t", session], text=True, capture_output=True, check=False)
        if existing.returncode == 0:
            attach = subprocess.run([tmux, "attach-session", "-t", session], text=True, capture_output=True, check=False)
            return int(attach.returncode), "attached existing TossQuant tmux session"
        return int(create.returncode), (create.stderr or create.stdout or "failed to create tmux session").strip()

    split = subprocess.run(
        [tmux, "split-window", "-t", target, "-v", "-l", str(height), "-c", str(cwd_path), hud_command],
        text=True,
        capture_output=True,
        check=False,
    )
    if split.returncode != 0:
        return int(split.returncode), (split.stderr or split.stdout or "failed to create HUD pane").strip()

    subprocess.run([tmux, "select-pane", "-t", f"{target}.0"], text=True, capture_output=True, check=False)
    attach = subprocess.run([tmux, "attach-session", "-t", session], text=True, capture_output=True, check=False)
    return int(attach.returncode), "TossQuant tmux runtime closed"

File: src/hud.ts
import { createHash } from 'node:crypto';
import { spawnSync } from 'node:child_process';
import { buildRuntimeSnapshot, readRuntimeSnapshot, recordRuntime, renderRuntimeLine, writeRuntimeSnapshot } from './runtime.ts';
import { hudColor } from './ui/hud.ts';

export const DEFAULT_SESSION = 'quantops';

export function sessionHash(source: string): string {
  return createHash('sha256').update(source).digest('hex').slice(0, 8);
}

export function defaultTmuxSession(env: NodeJS.ProcessEnv = process.env, cwd = process.cwd()): string {
  const source = env.QUANTOPS_SESSION || env.CODEX_SESSION_ID || env.OMX_SESSION_ID || env.OMX_SESSION || env.TMUX_PANE || cwd;
  return `${DEFAULT_SESSION}-${sessionHash(source)}`;
}

export function tmuxPath(): string | null {
  const result = spawnSync('sh', ['-lc', 'command -v tmux'], { encoding: 'utf8' });
  return result.status === 0 ? result.stdout.trim() : null;
}
export function tmuxInstallHint(): string { return 'install tmux with your OS package manager, e.g. apt install tmux, brew install tmux, or pacman -S tmux'; }
export function inTmux(): boolean { return Boolean(process.env.TMUX); }
export function shellCommand(parts: string[]): string { return parts.map((part) => `'${part.replaceAll("'", "'\\''")}'`).join(' '); }
export function managedTmuxSession(env: NodeJS.ProcessEnv = process.env): string | null {
  return env.QUANTOPS_TMUX_MANAGED === '1' && env.QUANTOPS_TMUX_SESSION ? env.QUANTOPS_TMUX_SESSION : null;
}
export function hudWatchCommand(base = 'data', interval = 1): string {
  return shellCommand([process.execPath, new URL('./cli.ts', import.meta.url).pathname, '--data-dir', base, 'hud', '--watch', '--interval', String(interval)]);
}
export function interactiveCommand(base = 'data', session?: string): string {
  const envPrefix = session ? ['env', 'QUANTOPS_TMUX_MANAGED=1', `QUANTOPS_TMUX_SESSION=${session}`] : [];
  return shellCommand([...envPrefix, process.execPath, new URL('./cli.ts', import.meta.url).pathname, '--no-tmux', '--data-dir', base]);
}

export function tmuxRuntimeOptions(session: string): string[][] {
  const target = `${session}:main`;
  return [
    ['set-option', '-t', session, 'mouse', 'on'],
    ['set-option', '-t', session, 'history-limit', '50000'],
    ['set-option', '-t', session, 'status-keys', 'vi'],
    ['set-option', '-t', session, 'renumber-windows', 'on'],
    ['set-window-option', '-t', target, 'mode-keys', 'vi'],
  ];
}

export function commandPaneTarget(session: string): string {
  return `${session}:main.0`;
}

export function hudPaneTarget(session: string): string {
  return `${session}:main.1`;
}

export function clampedHudHeight(requested: number, windowHeight?: number): number {
  const height = Math.max(1, Math.floor(requested));
  if (!Number.isFinite(windowHeight) || !windowHeight) return height;
  const maxHudHeight = Math.max(1, Math.floor(windowHeight) - 8);
  return Math.min(height, maxHudHeight);
}

function applyTmuxRuntimeOptions(tmux: string, session: string): void {
  for (const args of tmuxRuntimeOptions(session)) {
    spawnSync(tmux, args, { encoding: 'utf8' });
  }
}

function tmuxWindowHeight(tmux: string, target: string): number | undefined {
  const result = spawnSync(tmux, ['display-message', '-p', '-t', target, '#{window_height}'], { encoding: 'utf8' });
  const parsed = Number(result.stdout.trim());
  return Number.isFinite(parsed) ? parsed : undefined;
}

function selectCommandPane(tmux: string, session: string): void {
  const target = `${session}:main`;
  spawnSync(tmux, ['select-window', '-t', target], { encoding: 'utf8' });
  spawnSync(tmux, ['select-pane', '-t', commandPaneTarget(session)], { encoding: 'utf8' });
}

export function shutdownManagedTmuxRuntime(env: NodeJS.ProcessEnv = process.env): { code: number; message: string; skipped: boolean } {
  const session = managedTmuxSession(env);
  if (!session) return { code: 0, message: 'not a QuantOps-managed tmux runtime', skipped: true };
  const tmux = tmuxPath();
  if (!tmux) return { code: 127, message: `tmux not found in PATH; ${tmuxInstallHint()}`, skipped: true };
  const result = spawnSync(tmux, ['kill-session', '-t', session], { encoding: 'utf8' });
  return { code: result.status ?? 0, message: (result.stderr || result.stdout || `closed QuantOps tmux session ${session}`).trim(), skipped: false };
}

export function printHudOnce(base = 'data', mode = 'quant', lastAction = 'ready'): string {
  const line = renderRuntimeLine(recordRuntime({ base, mode, lastAction }));
  console.log(hudColor(line));
  return line;
}

export async function watchHud(base = 'data', interval = 1): Promise<never> {
  for (;;) {
    let snapshot = readRuntimeSnapshot(base);
    if (!snapshot) { snapshot = buildRuntimeSnapshot({ base }); writeRuntimeSnapshot(snapshot, base); }
    process.stdout.write(`\u001b[?25l\u001b[2J\u001b[H${hudColor(renderRuntimeLine(snapshot))}\u001b[0K`);
    await new Promise((resolve) => setTimeout(resolve, Math.max(interval, 0.2) * 1000));
  }
}

export function launchTmuxHud(base = 'data', height = 3, interval = 1): { code: number; message: string } {
  const tmux = tmuxPath();
  if (!tmux) return { code: 127, message: `tmux not found in PATH; ${tmuxInstallHint()}` };
  if (!inTmux()) return { code: 2, message: 'not inside a tmux session; run quantops with no arguments or start tmux first' };
  const result = spawnSync(tmux, ['split-window', '-v', '-l', String(Math.max(1, height)), hudWatchCommand(base, interval)], { encoding: 'utf8' });
  return { code: result.status ?? 1, message: (result.stderr || result.stdout || 'tmux HUD launched').trim() };
}

export function launchTmuxRuntime(base = 'data', session = DEFAULT_SESSION, height = 3, interval = 1, cwd = process.cwd()): { code: number; message: string } {
  const tmux = tmuxPath();
  if (!tmux) return { code: 127, message: `tmux not found in PATH; ${tmuxInstallHint()}` };
  if (inTmux()) return { code: 2, message: 'already inside tmux; use /hud tmux to add the QuantOps HUD pane' };
  const create = spawnSync(tmux, ['new-session', '-d', '-s', session, '-n', 'main', '-c', cwd, interactiveCommand(base, session)], { encoding: 'utf8' });
  if (create.status !== 0) {
    const exists = spawnSync(tmux, ['has-session', '-t', session], { encoding: 'utf8' });
    if (exists.status === 0) {
      applyTmuxRuntimeOptions(tmux, session);
      selectCommandPane(tmux, session);
      const attachExisting = spawnSync(tmux, ['attach-session', '-t', session], { encoding: 'utf8', stdio: 'inherit' });
      return { code: attachExisting.status ?? 0, message: 'attached existing QuantOps tmux session' };
    }
    return { code: create.status ?? 1, message: (create.stderr || create.stdout || 'failed to create tmux session').trim() };
  }
  applyTmuxRuntimeOptions(tmux, session);
  const target = `${session}:main`;
  spawnSync(tmux, ['select-pane', '-t', commandPaneTarget(session), '-T', 'QuantOps command'], { encoding: 'utf8' });
  const hudHeight = clampedHudHeight(height, tmuxWindowHeight(tmux, target));
  const split = spawnSync(tmux, ['split-window', '-t', target, '-v', '-l', String(hudHeight), '-c', cwd, hudWatchCommand(base, interval)], { encoding: 'utf8' });
  if (split.status !== 0) return { code: split.status ?? 1, message: (split.stderr || split.stdout || 'failed to create HUD pane').trim() };
  spawnSync(tmux, ['select-pane', '-t', hudPaneTarget(session), '-T', 'QuantOps HUD'], { encoding: 'utf8' });
  selectCommandPane(tmux, session);
  const attach = spawnSync(tmux, ['attach-session', '-t', session], { encoding: 'utf8', stdio: 'inherit' });
  return { code: attach.status ?? 0, message: 'QuantOps tmux runtime closed' };
}
